### PR TITLE
[LBSE] Improve RenderSVGResourcePattern

### PR DESCRIPTION
**Pull request by rwlbuis**
#### 9e8747f4819708b1c2cfa80ba0bf32a6042e9a2f
<pre>
[LBSE] Improve RenderSVGResourcePattern
<a href="https://bugs.webkit.org/show_bug.cgi?id=269184">https://bugs.webkit.org/show_bug.cgi?id=269184</a>

Reviewed by Nikolas Zimmermann.

Improve pattern resource implementation in the following ways:
- take legacy pattern resource calculations to fix custom/nested-pattern-boundingBoxModeContent.svg
- cache pattern tiles as image bitmaps (like legacy did). This allows pattern-nested-reference.html to not time out
- take care of correctly invalidating in case pattern contents change or a referencing client is added or removed

This PR relies on SVGGraphicsElement::invalidateResourceImageBuffersIfNeeded introduced in r277292 and adds
patterns as an SVG resource to invalidate for image buffer caching. The expectation is that this is the final
SVG resource that needs to be taken into account in SVGGraphicsElement::invalidateResourceImageBuffersIfNeeded.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/nested-pattern-boundingBoxModeContent-expected.txt: Added.
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
(WebCore::RenderSVGResourcePattern::buildPattern):
(WebCore::RenderSVGResourcePattern::prepareFillOperation):
(WebCore::RenderSVGResourcePattern::prepareStrokeOperation):
(WebCore::RenderSVGResourcePattern::createTileImage const):
(WebCore::RenderSVGResourcePattern::removeReferencingCSSClient):
(WebCore::clear2DRotation): Deleted.
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.h:
(WebCore::RenderSVGResourcePattern::invalidatePattern):
* Source/WebCore/svg/SVGGraphicsElement.cpp:
(WebCore::SVGGraphicsElement::invalidateResourceImageBuffersIfNeeded):
* Source/WebCore/svg/SVGPatternElement.cpp:
(WebCore::SVGPatternElement::childrenChanged):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::svgAttributeChanged):

Canonical link: <a href="https://commits.webkit.org/277306@main">https://commits.webkit.org/277306@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/86e1eed2da029f7a30f7a3645928263f203686d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/47272 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26452 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49923 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49956 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/43321 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49579 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/31875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23912 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38496 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47853 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23961 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40741 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19810 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/21392 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41903 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/5316 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43631 "Found 1 new API test failure: TestWebKitAPI.SleepDisabler.Close (failure)") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/42306 "Found 1 new test failure: inspector/cpu-profiler/tracking.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51831 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/22302 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18649 "Found 1 new test failure: inspector/cpu-profiler/tracking.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45790 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23577 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44802 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10419 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/24361 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/23295 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/svg/SVGElement.cpp [hunk 1]
```diff
@@ -284,15 +285,19 @@ Vector<WeakPtr<SVGResourceElementClient>> SVGElement::referencingCSSClients() co
     return copyToVector(m_svgRareData->referencingCSSClients());
 }
 
-void SVGElement::addReferencingCSSClient(SVGResourceElementClient& client)
+void SVGElement::addReferencingCSSClient(SVGResourceElementClient& client, const RenderElement& renderer)
```
Comment (nikolaszimmermann) [inline Source/WebCore/svg/SVGElement.cpp]: Why do we need to pass in "const RenderElement& renderer" here? where does it deviate from client->renderer()?

File: Source/WebCore/svg/SVGElement.cpp [hunk 2]
```diff
@@ -284,15 +285,19 @@ Vector<WeakPtr<SVGResourceElementClient>> SVGElement::referencingCSSClients() co
...
     ensureSVGRareData().addReferencingCSSClient(client);
 }
 
-void SVGElement::removeReferencingCSSClient(SVGResourceElementClient& client)
+void SVGElement::removeReferencingCSSClient(SVGResourceElementClient& client, const RenderElement& renderer)
 {
     if (!m_svgRareData)
         return;
+    if (CheckedPtr container = dynamicDowncast<RenderSVGResourceContainer>(this->renderer()))
+        container->removeReferencingCSSClient(renderer);
```
Comment (nikolaszimmermann) [inline Source/WebCore/svg/SVGElement.cpp]: Is it correct to omit this call if m_svgRareDate is null? This looks unrelated.

**Review by nikolaszimmermann — COMMENTED** [review]
Left some comments, that also apply to the mask patch :-)

File: Source/WebCore/svg/SVGElement.cpp (hunk 2 above)
Comment (rwlbuis) [reply Source/WebCore/svg/SVGElement.cpp]: If there is no rare data, there are no clients that can be removed.

File: Source/WebCore/svg/SVGElement.cpp (hunk 1 above)
Comment (rwlbuis) [reply Source/WebCore/svg/SVGElement.cpp]: There is no such method in SVGResourceElementClient. Would you prefer a virtual method for that in SVGResourceElementClient?

Comment (nikolaszimmermann) [reply Source/WebCore/svg/SVGElement.cpp]: No, I was wondering why you need to pass in the renderer at all, if it is client.renderer() that you pass around?

Comment (rwlbuis) [reply Source/WebCore/svg/SVGElement.cpp]: This should be addressed by the masker PR, indeed it seems cleaner to just ask the client.

Comment (rwlbuis) [reply Source/WebCore/svg/SVGElement.cpp]: Maybe you were looking at an older change?

**Review by nikolaszimmermann — APPROVED** [review]
Nice how small this is now :-)
